java/trace/src/main/scala/org/typelevel/otel4s/java/trace/TracerImpl.scala

```diff
@@ -73,4 +80,129 @@ private[java] class TracerImpl[F[_]: Sync](
         rootScope(fa)
     }
   }
+
+  def translate[G[_]](fk: F ~> G, gk: G ~> F): Tracer[G] =
+    new Tracer[G] {
+      private implicit val G: Sync[G] =
+        TracerImpl.liftSync[F, G](Sync[F], fk, gk)
+
+      private val traceScope: TraceScope[G] =
+        new TraceScope[G] {
+          def root: G[Scope.Root] = fk(scope.root)
+
+          def current: G[Scope] = fk(scope.current)
+
+          def makeScope(span: JSpan): G[G ~> G] =
+            fk(scope.makeScope(span)).map { transform =>
+              new (G ~> G) {
+                def apply[A](fa: G[A]): G[A] =
+                  fk(transform.apply(gk(fa)))
+              }
+            }
+
+          def rootScope: G[G ~> G] =
+            fk(scope.rootScope).map { transform =>
+              new (G ~> G) {
+                def apply[A](fa: G[A]): G[A] =
+                  fk(transform.apply(gk(fa)))
+              }
+            }
+
+          def noopScope: G ~> G =
+            new (G ~> G) {
+              def apply[A](fa: G[A]): G[A] =
+                fk(scope.noopScope(gk(fa)))
+            }
+        }
+
+      private val spanRunner: SpanRunner[G, Span[G]] =
+        SpanRunner.span(traceScope)
+
+      val meta: Tracer.Meta[G] =
+        new Tracer.Meta[G] {
+          val noopSpanBuilder: SpanBuilder.Aux[G, Span[G]] =
+            SpanBuilder.noop(Span.Backend.noop[G])
+          val isEnabled: Boolean = self.meta.isEnabled
+          val unit: G[Unit] = MonadCancelThrow[G].unit
+        }
+
+      def currentSpanContext: G[Option[SpanContext]] =
+        fk(self.currentSpanContext)
+
+      def spanBuilder(name: String): Aux[G, Span[G]] =
+        new SpanBuilderImpl[G, Span[G]](
+          jTracer,
+          name,
+          traceScope,
+          spanRunner
+        )
+
+      def childScope[A](parent: SpanContext)(fa: G[A]): G[A] =
+        fk(self.childScope(parent)(gk(fa)))
+
+      def joinOrRoot[A, C: TextMapGetter](carrier: C)(fa: G[A]): G[A] =
+        fk(self.joinOrRoot(carrier)(gk(fa)))
+
+      def rootScope[A](fa: G[A]): G[A] =
+        fk(self.rootScope(gk(fa)))
+
+      def noopScope[A](fa: G[A]): G[A] =
+        fk(self.noopScope(gk(fa)))
+
+      def translate[Q[_]](fk1: G ~> Q, gk1: Q ~> G): Tracer[Q] =
+        self.translate[Q](fk.andThen(fk1), gk1.andThen(gk))
+    }
+
+}
+
+object TracerImpl {
+
+  private def liftSync[F[_], G[_]](
```

Comment:
Nice! We might consider upstreaming these methods to Cats Effect is they are generally useful.

Reply:
That's what I was thinking about too